### PR TITLE
Expand libstd struct case misspelling diagnostics

### LINKED ISSUE #72641: Suggest importable type that differs only in capitalization
```rust
struct S {
    m: Hashmap<String, ()>,
}
```

```console
error[E0412]: cannot find type `Hashmap` in this scope
 --> src/main.rs:2:8
  |
2 |     m: Hashmap<String, ()>,
  |        ^^^^^^^ not found in this scope
```

If the code had been written with `HashMap` instead of `Hashmap`, we would get the following extremely likely correct suggestion:

```console
error[E0412]: cannot find type `HashMap` in this scope
 --> src/main.rs:2:8
  |
2 |     m: HashMap<String, ()>,
  |        ^^^^^^^ not found in this scope
  |
help: consider importing one of these items
  |
1 | use std::collections::HashMap;
  |
1 | use std::collections::hash_map::HashMap;
  |
```

`Hashmap` is a fairly common capitalization in other languages (https://grep.app/search?q=Hashmap&case=true) so it would be good to provide an appropriate suggestion in this case.

(For whatever reason this doesn't trigger #72640.)

### PR DESCRIPTION
**Pull request by chrissimpkins**
Closes #72641 

This PR adds an additional fallback `librustc_resolve` diagnostic with libstd struct suggestions for case misspelling errors under the following conditions:

- libstd struct with correct spelling was not brought into scope (they are already handled with a different diagnostics approach)
- struct is not imported in the Rust prelude (they are already handled with a different diagnostics approach)
- not a primitive type (they are already handled with a different diagnostics approach)
- struct is not nightly-only
- case-insensitive struct name does not match a macro name (the macro name suggestion has precedence with the current approach here and these suggestions are not displayed)
- struct is not deprecated (e.g., `std::str::LinesAny`)
- struct is not an error type (arbitrary decision)

Behavior: A user who enters a struct name with a libstd struct case misspelling for a type that is not in scope will receive the matching libstd struct(s) in a new suggestion.  This approach was originally requested in #72641 for `Hashmap` -> `HashMap` only.  These changes broaden the scope of that report to cover many/most libstd structs with any case misspelling error, and support multiple struct suggestions when there are multiple case-insensitive name matches across the standard library. This is only used when the Levenshtein algorithm does not identify a typo candidate match.

Before:

```
error[E0412]: cannot find type `Hashmap` in this scope
 --> src/main.rs:4:9
  |
4 |     hm: Hashmap<String, ()>,
  |         ^^^^^^^ not found in this scope

error: aborting due to previous error

For more information about this error, try `rustc --explain E0412`.
```

After:

```
error[E0412]: cannot find type `Hashmap` in this scope
 --> src/main.rs:4:9
  |
4 |     hm: Hashmap<String, ()>,
  |         ^^^^^^^
  |         |
  |         not found in this scope
  |         help: found a std struct with a similar name: `std::collections::HashMap`

error: aborting due to previous error
```

### CODE REVIEW

**Comment by chrissimpkins** [comment]
@matthewjasper I think that this is ready for review after some edits to address issues that came up in other UI tests following the initial push.  CI is failing on some UI tests that I did not add.  Assuming that these changes are acceptable, we can discuss how to approach the `Drain` issue.  The only other outstanding issues are the `HashMap` UI tests and IMO these are OK.  I can `--bless` them if you agree.  Will wait on doing anything further until you have a chance to review.  Let me know if there is anything that you need from me.  Thanks!

**Comment by euclio** [comment]
[Generally we try to avoid "did you mean" in diagnostic suggestions](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/struct.Diagnostic.html#method.span_suggestion). Instead, I would word this to indicate what we're suggesting and why we're suggesting it. Something like "the standard library contains a \<item type\> with a similar name".

**Comment by chrissimpkins** [comment]
> [Generally we try to avoid "did you mean" in diagnostic suggestions](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/struct.Diagnostic.html#method.span_suggestion). Instead, I would word this to indicate what we're suggesting and why we're suggesting it. Something like "the standard library contains a <item type> with a similar name".

[We have some work to do](https://github.com/rust-lang/rust/search?q=%22did+you+mean%22)... ;)

Happy to change it.

File: src/test/ui/libstd-case-typo/alloc.stderr [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+error[E0412]: cannot find type `system` in this scope
+  --> $DIR/alloc.rs:6:20
+   |
+LL | fn test_system(_x: system){}
+   |                    ^^^^^^ not found in this scope
+   |
+help: did you mean `std::alloc::System`?
+   |
+LL | fn test_system(_x: std::alloc::System){}
```
Comment (Patryk27) [inline src/test/ui/libstd-case-typo/alloc.stderr:20]: IMO repeating the whole expression makes the hint look cluttered, and - at least for me - quite hard to follow; what do you say we just print the name of the suggested type?
```
error[E0412]: cannot find type `system` in this scope
  --> $DIR/alloc.rs:6:20
   |
LL | fn test_system(_x: system){}
   |                    ^^^^^^ help: a struct with a similar name exists: `std::alloc::System`
   |
```
For reference, you already get a similar message for full paths:
```rust
struct Foo {
    x: std::collections::HashMaP<(), ()>,
}
```
```
error[E0412]: cannot find type `HashMaP` in module `std::collections`
   --> src/lib.rs:4:26
    |
4   |       x: std::collections::HashMaP<(), ()>,
    |                            ^^^^^^^ help: a struct with a similar name exists: `HashMap`
```

Comment (chrissimpkins) [reply src/test/ui/libstd-case-typo/alloc.stderr:20]: The error label on that span is more generally used and is not being defined here.  We will continue to get the "not found in this scope" string without edits to that area of the diagnostics.  The libstd suggestions in this PR are being appended to the error diagnostics that occur further up the type check chain.

I can remove the separate suggestion block.  I intentionally used the verbose form of the suggestion because I find the formatting preferable to the default version.  Here is what I get with the change:

```
error[E0412]: cannot find type `Hashmap` in this scope
 --> src/main.rs:4:9
  |
4 |     hm: Hashmap<String, ()>,
  |         ^^^^^^^
  |         |
  |         not found in this scope
  |         help: found a libstd struct with a similar name `std::collections::HashMap`?: `std::collections::HashMap`

error: aborting due to previous error
```

The suggestion string (here the full struct path) is repeated on the same line following a colon. We could eliminate the suggestion altogether and just format it as a label but IIUC this eliminates the ability for autofixes by tools.

Thoughts?

cc: @euclio with the suggestion string change above

Comment (chrissimpkins) [reply src/test/ui/libstd-case-typo/alloc.stderr:20]: The error label on the span ("not found in this scope") is added in a more general check further up the diagnostics chain.  We can't eliminate it without modifying that source.

I intentionally used the verbose version of the suggestion because I tend to like the formatting.  It is set off with a suggestion in a different color and I think that this makes it distinct from the message about what the error is.  I don't have strong opinions about this though.  Here is how it appears with the default suggestion format and a change in wording to address your and @euclio concerns:

```
error[E0412]: cannot find type `Hashmap` in this scope
 --> src/main.rs:4:9
  |
4 |     hm: Hashmap<String, ()>,
  |         ^^^^^^^
  |         |
  |         not found in this scope
  |         help: found a libstd struct with a similar name: `std::collections::HashMap`

error: aborting due to previous error
```

Comment (chrissimpkins) [reply src/test/ui/libstd-case-typo/alloc.stderr:20]: Thoughts?

Comment (chrissimpkins) [reply src/test/ui/libstd-case-typo/alloc.stderr]: Pushed these changes in https://github.com/rust-lang/rust/pull/72988/commits/bf2f78eaab4b9a9044011921bb2f31d18b8bb5ab so that you can see all of the UI test stderr strings with the new format.

Comment (Patryk27) [reply src/test/ui/libstd-case-typo/alloc.stderr]: The new version looks really neat! :smile: 

I'd suggest two adjustments:
- `found libstd structs with a similar name` -> `found libstd structs with similar names` (since `names` refers to those newly-found structs)
- `libstd` -> `std` (I think it's more ubiquitous & newbie-friendly term than `libstd`)

Comment (chrissimpkins) [reply src/test/ui/libstd-case-typo/alloc.stderr]: Done!

https://github.com/rust-lang/rust/pull/72988/commits/9441340d0be1922b047f82dbcde516ad0d52cac7

Comment (chrissimpkins) [reply src/test/ui/libstd-case-typo/alloc.stderr]: Here is where we are:

```
error[E0412]: cannot find type `Hashmap` in this scope
 --> src/main.rs:4:9
  |
4 |     hm: Hashmap<String, ()>,
  |         ^^^^^^^
  |         |
  |         not found in this scope
  |         help: found a std struct with a similar name: `std::collections::HashMap`

error: aborting due to previous error
```

Comment (Patryk27) [reply src/test/ui/libstd-case-typo/alloc.stderr]: wonderful, thanks :-)

Comment (chrissimpkins) [reply src/test/ui/libstd-case-typo/alloc.stderr]: Thanks for your feedback.  I really appreciate it!

**Comment by chrissimpkins** [comment]
@davidtwco David is this something that you would be willing to review?  This is related to https://github.com/rust-lang/rust/pull/73023 that you recently reviewed.

**Comment by chrissimpkins** [comment]
> r? @davidtwco
> 
> I'll take a look at this shortly.

Tyvm!

File: src/librustc_resolve/late/diagnostics.rs [hunk 2]
```diff
@@ -353,10 +353,252 @@ impl<'a> LateResolutionVisitor<'a, '_, '_> {
...
+
         (err, candidates)
     }
 
+    /// Get a case-insensitive match with standard library
+    /// structs that are *not imported* in the prelude.
+    /// This is used for type checking diagnostics in cases when
+    /// the type is not in scope and the name includes case
+    /// misspelling (e.g., `Hashmap`, not `HashMap`).
+    fn get_case_insensitive_libstd_structs_matches(&self, needle: &str) -> Vec<String> {
```
Comment (davidtwco) [inline src/librustc_resolve/late/diagnostics.rs:398]: I think we should take a different approach - I'd rather we avoid hard-coding `std` type names in the compiler (the compiler shouldn't need to be updated when `std` is changed for this diagnostic to continue to work) and I think we can make this more general, so it works for more than just `std`.

---

As you note in the PR description, we do Levenshtein matching on things that are already in-scope and suggest fixes in that case. We also already have suggestions for things that aren't in scope but where the name matches exactly.

The snippet below is where the compiler decides to make "did you mean to import this" suggestions:

https://github.com/rust-lang/rust/blob/50c0192c64241d723066add22c53d472e2b9cba9/src/librustc_resolve/diagnostics.rs#L660-L693

On line 660 is where it checks for an exact match - I'd experiment with doing a exact match *or* checking for a Levenshtein distance from that name that's under some threshold (look into `lev_distance` and `find_best_match_for_name` to see what can be re-used). I'd hope that would detect cases like this where we can import a name that's very similar. Alternatively, we could just change the comparison to compare the names without considering the case (as I understand it, that would be equivalent to what you're doing here?).

What we might also want to do is thread this information through in some way - e.g. changing `ImportSuggestion` to be an enum with `Exact` and `Almost` (you can come up with a better name than this) variants so that we can make the diagnostic separate these two types of suggestions:

```
error[E0412]: cannot find type `Hashmap` in this scope
 --> src/main.rs:2:8
  |
2 |     m: Hashmap<String, ()>,
  |        ^^^^^^^ not found in this scope
  |
help: consider importing one of these items
  |
1 | use some::made:up::module::Hashmap;
  |
help: consider importing one of these similarly named items
  |
1 | use std::collections::HashMap;
  |
```

This should be easier to maintain because we don't have to keep a list of types like you've got here, and it'd work for non-`std` types too.

File: src/test/ui/libstd-case-typo/char.rs [hunk 3]
```diff
@@ -0,0 +1,17 @@
+// checks case typos with libstd::char structs
+fn main(){}
+
+fn test_du16(_x: DecodeUTF16<()>){}
+//~^ ERROR: cannot find type `DecodeUTF16` in this scope
+
+fn test_edflt(_x: Escapedefault){}
```
Comment (davidtwco) [inline src/test/ui/libstd-case-typo/char.rs:7]: I'd combine all of these tests into a single file - we're not gaining anything by having them separate - if we change the approach, it's very likely that if one fails then they all would and so having different tests just means we have more tests to run (likely to be marginally slower).

**Review by davidtwco — CHANGES_REQUESTED** [review]

File: src/librustc_resolve/late/diagnostics.rs (hunk 2 above)
Comment (chrissimpkins) [reply src/librustc_resolve/late/diagnostics.rs:398]: Tyvm David!  I will look into it this week.  I really appreciate the review and will be in touch when I have updates.

Comment (estebank) [reply src/librustc_resolve/late/diagnostics.rs:398]: Small nitpick on wording: `help: consider importing one of these similarly named items` → `help: you might have meant to use {one of }the following similarly named item{s}`.

Comment (davidtwco) [reply src/librustc_resolve/late/diagnostics.rs:398]: @chrissimpkins How are you getting with this? Feel free to open a work-in-progress PR if you’re unsure how to proceed or want some early feedback!

Comment (chrissimpkins) [reply src/librustc_resolve/late/diagnostics.rs:398]: Sorry for the delay! Busy week last week. I will get back to this during the week and let you know if I run into any issues.

**Comment by Muirrum** [comment]
@chrissimpkins This is a triage bump.

**Comment by chrissimpkins** [comment]
@Muirrum Thanks for the reminder Owen.  I don't know that any of this source will be used in what is required here based on the review.  I'll close this and open a new PR when time permits me to dig into it again.
